// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2019 Egor Pugin <egor.pugin@gmail.com>

syntax = "proto3";

package sw.api.build;

//import "google/protobuf/empty.proto";

message IOStream {
    string file = 1;
    string text = 2;
    //bool inherit = 3;
}

message Command {
    uint64 id = 1; // to track deps

    repeated string arguments = 5;
    string working_directory = 6;
    map<string, string> environment = 7;

    IOStream in = 11;
    IOStream out = 12;
    IOStream err = 13;

    repeated uint64 dependencies = 21;
}

message Commands {
    repeated Command commands = 1;
}

// CommandResponse?
message CommandResult {
    int64 exit_code = 1;
    // pid?

    string out = 9;
    string err = 10;
}

// add execution plan?

service DistributedBuildService {
    rpc ExecuteCommand(Command) returns (CommandResult);
}
